File: sdks/python-admin/aegis_db_admin/shield.py
"""
Aegis-DB Admin SDK - Shield Service

Security shield management: status, events, IP blocking, allowlisting,
policy, IP reputation, and threat feeds.

@version 1.0.0
@author AutomataNexus Development Team
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, TYPE_CHECKING

from .types import (
    ShieldStatus,
    ShieldStats,
    SecurityEvent,
    BlockedIP,
    AllowlistEntry,
    ShieldPolicy,
    IPReputation,
    ThreatFeedEntry,
)

if TYPE_CHECKING:
    from .client import AegisAdmin


class ShieldService:
    """Manage the security shield."""

    def __init__(self, admin: AegisAdmin) -> None:
        self._admin = admin

    # =========================================================================
    # Status / Stats
    # =========================================================================

    async def get_status(self) -> ShieldStatus:
        """Get shield status.

        Returns:
            ShieldStatus object.
        """
        data = await self._admin._request("GET", "/api/v1/shield/status")
        return ShieldStatus(
            enabled=data.get("enabled", False),
            mode=data.get("mode"),
            rules_count=data.get("rules_count", 0),
            _raw=data,
        )

    async def get_stats(self) -> ShieldStats:
        """Get shield statistics.

        Returns:
            ShieldStats object.
        """
        data = await self._admin._request("GET", "/api/v1/shield/stats")
        return ShieldStats(
            total_requests=data.get("total_requests", 0),
            blocked_requests=data.get("blocked_requests", 0),
            threats_detected=data.get("threats_detected", 0),
            _raw=data,
        )

    # =========================================================================
    # Security Events
    # =========================================================================

    async def get_events(self) -> List[SecurityEvent]:
        """Get security events.

        Returns:
            List of SecurityEvent objects.
        """
        data = await self._admin._request("GET", "/api/v1/shield/events")
        events_list = data if isinstance(data, list) else data.get("events", [])
        return [_parse_security_event(e) for e in events_list]

    # =========================================================================
    # Blocked IPs
    # =========================================================================

    async def list_blocked(self) -> List[BlockedIP]:
        """List blocked IPs.

        Returns:
            List of BlockedIP objects.
        """
        data = await self._admin._request("GET", "/api/v1/shield/blocked")
        blocked_list = data if isinstance(data, list) else data.get("blocked", [])
        return [
            BlockedIP(
                ip=b.get("ip"),
                reason=b.get("reason"),
                blocked_at=b.get("blocked_at"),
                expires_at=b.get("expires_at"),
                _raw=b,
            )
            for b in blocked_list
        ]

    async def block_ip(
        self,
        ip: str,
        *,
        reason: Optional[str] = None,
        duration: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Block an IP address.

        Args:
            ip: The IP address to block.
            reason: Optional reason for blocking.
            duration: Optional duration (e.g. '24h', '7d').

        Returns:
            Server response dict.
        """
        payload: Dict[str, Any] = {"ip": ip}
        if reason is not None:
            payload["reason"] = reason
        if duration is not None:
            payload["duration"] = duration
        return await self._admin._request("POST", "/api/v1/shield/blocked", payload)

    async def unblock_ip(self, ip: str) -> Dict[str, Any]:
        """Unblock an IP address.

        Args:
            ip: The IP address to unblock.

        Returns:
            Server response dict.
        """
        return await self._admin._request("DELETE", f"/api/v1/shield/blocked/{ip}")

    # =========================================================================
    # Allowlist
    # =========================================================================

    async def get_allowlist(self) -> List[AllowlistEntry]:
        """Get the IP allowlist.

        Returns:
            List of AllowlistEntry objects.
        """
        data = await self._admin._request("GET", "/api/v1/shield/allowlist")
        allow_list = data if isinstance(data, list) else data.get("allowlist", [])
        return [
            AllowlistEntry(
                ip=a.get("ip"),
                added_at=a.get("added_at"),
                description=a.get("description"),
                _raw=a,
            )
            for a in allow_list
        ]

    async def add_to_allowlist(
        self,
        ip: str,
        *,
        description: Optional[str] = None,
    ) -> Dict[str, Any]:
        """Add an IP to the allowlist.

        Args:
            ip: The IP address to allow.
            description: Optional description.

        Returns:
            Server response dict.
        """
        payload: Dict[str, Any] = {"ip": ip}
        if description is not None:
            payload["description"] = description
        return await self._admin._request("POST", "/api/v1/shield/allowlist", payload)

    async def remove_from_allowlist(self, ip: str) -> Dict[str, Any]:
        """Remove an IP from the allowlist.

        Args:
            ip: The IP address to remove.

        Returns:
            Server response dict.
        """
        return await self._admin._request("DELETE", f"/api/v1/shield/allowlist/{ip}")

    # =========================================================================
    # Policy
    # =========================================================================

    async def get_policy(self) -> ShieldPolicy:
        """Get the shield security policy.

        Returns:
            ShieldPolicy object.
        """
        data = await self._admin._request("GET", "/api/v1/shield/policy")
        return ShieldPolicy(_raw=data)

    async def update_policy(self, policy: Dict[str, Any]) -> ShieldPolicy:
        """Update the shield security policy.

        Args:
            policy: Policy configuration dict.

        Returns:
            Updated ShieldPolicy object.
        """
        data = await self._admin._request("PUT", "/api/v1/shield/policy", policy)
        return ShieldPolicy(_raw=data)

    # =========================================================================
    # IP Reputation
    # =========================================================================

    async def get_ip_reputation(self, ip: str) -> IPReputation:
        """Get reputation information for an IP address.

        Args:
            ip: The IP address to look up.

        Returns:
            IPReputation object.
        """
        data = await self._admin._request("GET", f"/api/v1/shield/ip/{ip}")
        return IPReputation(
            ip=data.get("ip", ip),
            score=data.get("score"),
            classification=data.get("classification"),
            _raw=data,
        )

    # =========================================================================
    # Threat Feed
    # =========================================================================

    async def get_threat_feed(self) -> List[ThreatFeedEntry]:
        """Get the threat feed.

        Returns:
            List of ThreatFeedEntry objects.
        """
        data = await self._admin._request("GET", "/api/v1/shield/feed")
        feed_list = data if isinstance(data, list) else data.get("feed", [])
        return [ThreatFeedEntry(_raw=f) for f in feed_list]


# =============================================================================
# Helpers
# =============================================================================

def _parse_security_event(data: Dict[str, Any]) -> SecurityEvent:
    return SecurityEvent(
        id=data.get("id"),
        event_type=data.get("event_type"),
        severity=data.get("severity"),
        source_ip=data.get("source_ip"),
        message=data.get("message"),
        timestamp=data.get("timestamp"),
        _raw=data,
    )

File: sdks/javascript-admin/src/shield.ts
/**
 * Shield service for security monitoring, IP blocking, and threat management.
 */

import type { RequestFn } from './index';

// ============================================================================
// Types
// ============================================================================

export interface ShieldStatus {
  enabled: boolean;
  mode: string;
  patterns_loaded: number;
  blocked_ips: number;
  allowlisted_ips: number;
  events_total: number;
}

export interface ShieldStats {
  total_requests: number;
  blocked_requests: number;
  sql_injection_attempts: number;
  xss_attempts: number;
  rate_limited: number;
  suspicious_ips: number;
  threats_detected: number;
  period?: string;
}

export interface ShieldEvent {
  id: string;
  timestamp: string;
  event_type: string;
  severity: string;
  source_ip: string;
  description: string;
  details?: Record<string, unknown>;
  blocked: boolean;
}

export interface BlockedIP {
  ip: string;
  reason: string;
  blocked_at: string;
  expires_at?: string;
  auto_blocked: boolean;
}

export interface AllowlistEntry {
  ip: string;
  description?: string;
  added_at: string;
  added_by?: string;
}

export interface BlockIPRequest {
  ip: string;
  reason: string;
  duration_seconds?: number;
}

export interface AllowlistRequest {
  ip: string;
  description?: string;
}

export interface ShieldPolicy {
  sql_injection_detection: boolean;
  xss_detection: boolean;
  rate_limiting: boolean;
  ip_reputation: boolean;
  auto_block: boolean;
  auto_block_threshold: number;
  max_request_size_bytes?: number;
  request_rate_limit?: number;
  login_rate_limit?: number;
  [key: string]: unknown;
}

export interface IPReputation {
  ip: string;
  score: number;
  risk_level: string;
  total_requests: number;
  blocked_requests: number;
  last_seen: string;
  flags: string[];
}

export interface ThreatFeed {
  entries: ThreatFeedEntry[];
  last_updated: string;
  source?: string;
}

export interface ThreatFeedEntry {
  ip: string;
  threat_type: string;
  confidence: number;
  first_seen: string;
  last_seen: string;
}

// ============================================================================
// Service
// ============================================================================

export class ShieldService {
  constructor(private request: RequestFn) {}

  /** Get shield status. */
  async getStatus(): Promise<ShieldStatus> {
    return this.request<ShieldStatus>('GET', '/api/v1/shield/status');
  }

  /** Get shield statistics. */
  async getStats(): Promise<ShieldStats> {
    return this.request<ShieldStats>('GET', '/api/v1/shield/stats');
  }

  /** Get security events. */
  async getEvents(): Promise<ShieldEvent[]> {
    return this.request<ShieldEvent[]>('GET', '/api/v1/shield/events');
  }

  /** List blocked IPs. */
  async listBlocked(): Promise<BlockedIP[]> {
    return this.request<BlockedIP[]>('GET', '/api/v1/shield/blocked');
  }

  /** Block an IP address. */
  async blockIP(ip: string, reason: string, durationSeconds?: number): Promise<BlockedIP> {
    return this.request<BlockedIP>('POST', '/api/v1/shield/blocked', {
      ip,
      reason,
      duration_seconds: durationSeconds,
    });
  }

  /** Unblock an IP address. */
  async unblockIP(ip: string): Promise<void> {
    await this.request<void>('DELETE', `/api/v1/shield/blocked/${encodeURIComponent(ip)}`);
  }

  /** List allowlisted IPs. */
  async listAllowlist(): Promise<AllowlistEntry[]> {
    return this.request<AllowlistEntry[]>('GET', '/api/v1/shield/allowlist');
  }

  /** Add an IP to the allowlist. */
  async addToAllowlist(ip: string, description?: string): Promise<AllowlistEntry> {
    return this.request<AllowlistEntry>('POST', '/api/v1/shield/allowlist', {
      ip,
      description,
    });
  }

  /** Remove an IP from the allowlist. */
  async removeFromAllowlist(ip: string): Promise<void> {
    await this.request<void>('DELETE', `/api/v1/shield/allowlist/${encodeURIComponent(ip)}`);
  }

  /** Get the current shield security policy. */
  async getPolicy(): Promise<ShieldPolicy> {
    return this.request<ShieldPolicy>('GET', '/api/v1/shield/policy');
  }

  /** Update the shield security policy. */
  async updatePolicy(policy: Partial<ShieldPolicy>): Promise<ShieldPolicy> {
    return this.request<ShieldPolicy>('PUT', '/api/v1/shield/policy', policy);
  }

  /** Get IP reputation details. */
  async getIPReputation(ip: string): Promise<IPReputation> {
    return this.request<IPReputation>('GET', `/api/v1/shield/ip/${encodeURIComponent(ip)}`);
  }

  /** Get the threat feed. */
  async getThreatFeed(): Promise<ThreatFeed> {
    return this.request<ThreatFeed>('GET', '/api/v1/shield/feed');
  }
}
